zio-crypto/shared/src/main/scala/zio/crypto/hashing/Hashing.scala

```diff
@@ -0,0 +1,115 @@
+package zio.crypto.hashing
+
+import zio._
+import zio.crypto.ByteHelpers
+
+import java.nio.charset.Charset
+import java.security.{ MessageDigest => JMessageDigest }
+
+sealed class HashAlgorithm(val name: String)
+
+object HashAlgorithm {
+  case object SHA256 extends HashAlgorithm("SHA-256")
+  case object SHA512 extends HashAlgorithm("SHA-512")
+
+  // DO NOT USE IN A SECURE CONTEXT
+  case object MD5 extends HashAlgorithm("MD5")
+  // DO NOT USE IN A SECURE CONTEXT
+  case object SHA1 extends HashAlgorithm("SHA-1")
+}
+
+case class MessageDigest[T](value: T) extends AnyVal
+
+object Hashing {
+
+  type Hashing = Has[Hashing.Service]
+
+  trait Service {
+    def hash(m: Array[Byte], alg: HashAlgorithm): Task[MessageDigest[Array[Byte]]]
+    def verify(m: Array[Byte], digest: MessageDigest[Array[Byte]], alg: HashAlgorithm): Task[Boolean]
```

Comment:
If a function is deterministic and free of observable side-effects, then it should not return the value wrapped in `Task`. 

As far as I can see, all of these can be pure functions. Right?

Reply:
`hash` can throw `NoSuchAlgorithmException`, which is why I had wrapped it in `Task`. `verify` calls `hash` on `m` to check against `digest`, which is why `verify` is a `Task` as well. Perhaps that is an acceptable and unlikely failure? The remaining possible failure is in the `String` versions of these functions, where the supplied base 64 encoded string may be invalid.